feat: add GPS UART module with NMEA parsing and checksum validation

yousefyasser opened the pull request:
- Added checksum validation for incoming NMEA strings
- Added parsing of NMEA strings of type GPRMC to extract needed information, such as: latitude and longitude

Co-authored-by: @RasheedAtia
Co-authored-by: @seifAbbas

yousefyasser committed:
feat: add GPS UART module with NMEA parsing and checksum validation

Co-authored-by: NightsThunder <saifomaster4@gmail.com>
Co-authored-by: RasheedAtia <rasheedatia@hotmail.com>

yousefyasser commented:
Hello @lurch, ok I will add the main function right now

yousefyasser committed:
feat: implement main function for GPS data processing and output

yousefyasser committed:
doc: add detailed documentation for GPS UART driver functions

lurch commented on uart/gps_uart/gps_uart.c:
```diff
@@ -0,0 +1,250 @@
+
+    // sufficient waiting time required between each gps reading and the next one
+    sleep_ms(30 * 1000);
+    tight_loop_contents();
```
No need for `tight_loop_contents` here - it's only intended for busy-wait-loops in which nothing else is happening.

lurch commented:
> Hello @lurch, ok I will add the main function right now

Thanks! Could you also please clarify which licence this code is released under?

yousefyasser committed:
refactor: remove unnecessary tight_loop_contents call

lurch commented:
> Sure, this code is released under the MIT license

A comment in the source-code saying that (for the benefit of future readers) would be great.

yousefyasser committed:
doc: add SPDX license identifier to GPS UART driver file

peterharperuk commented:
What hardware are you using to test this? I've only seen i2c gps modules so far (I haven't looked too hard).

peterharperuk commented on uart/gps_uart/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+# Link to pico_stdlib (gpio, time, etc. functions)
+target_link_libraries(gps_uart pico_stdlib hardware_uart)
+
+# Enable usb output, disable uart output
```
misleading comment

peterharperuk commented on uart/gps_uart/gps_uart.c:
```diff
@@ -0,0 +1,251 @@
+    char nmea_buffer[MAX_NMEA_LENGTH];
+    int chars_read = 0;
+
+    while (uart_is_readable(UART_ID) && chars_read < MAX_NMEA_LENGTH - 1)
```
This function doesn't handle uart_is_readable being false before the end of the line

peterharperuk commented on uart/gps_uart/gps_uart.c:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    // Check if we have enough tokens and data is valid
+    if (token_count >= 12 && strcmp(tokens[2], "A") == 0)
```
This fails for this string "$GPRMC,193045.000,A,5214.2340,N,00008.7338,E,0.00,0.00,130225,,,D*66"

peterharperuk reviewed (changes requested):
Doesn't work for my gps device https://shop.pimoroni.com/products/adafruit-ultimate-gps-breakout?variant=288854706

peterharperuk commented on uart/gps_uart/gps_uart.c:
```diff
@@ -156,4 +157,35 @@ void process_gps_data(GPSData *gps_data)
+  uart_gps_init();
+
+  // recommended cold start waiting time (could vary based on the gps module)
+  sleep_ms(7 * 60 * 1000);
```
sleeping for 7 minutes is madness. This shouldn't be needed

peterharperuk commented on uart/gps_uart/gps_uart.c:
```diff
@@ -156,4 +157,35 @@ void process_gps_data(GPSData *gps_data)
+    }
+
+    // sufficient waiting time required between each gps reading and the next one
+    sleep_ms(30 * 1000);
```
Why sleep at all?

peterharperuk committed:
Add GPS example to readme

peterharperuk committed:
Add gps_uart to the build

peterharperuk committed:
Improve parsing

This was failing to read whole lines for me. And for some reason my
device is putting zeros on the head of strings.

peterharperuk committed:
Remove sleeps

This seem unnecessary.

peterharperuk committed:
Remove debugging output.

Make the results clearer.

lurch commented on uart/gps_uart/gps_uart.c:
```diff
@@ -0,0 +1,260 @@
+// please refer to the datasheet
+#define UART_ID uart1
+#define BAUD_RATE 9600
+#define UART_TX_PIN 4
+#define UART_RX_PIN 5
```
@peterharperuk Is it worth putting these in the pico_binary_info thingy?
